plugins/hls-overloaded-record-dot-plugin/src/Ide/Plugin/OverloadedRecordDot.hs

```diff
@@ -0,0 +1,240 @@
+                                                 LHsExpr, Outputable, Pass (..),
+                                                 RealSrcSpan, getLoc, hs_valds,
+                                                 pattern RealSrcSpan, unLoc)
+import           Development.IDE.GHC.Util       (getExtensions, printOutputable)
+import           Development.IDE.Graph          (RuleResult)
+import           Development.IDE.Graph.Classes  (Hashable, NFData (rnf))
+import           Development.IDE.Spans.Pragmas  (NextPragmaInfo (..),
+                                                 getFirstPragma,
+                                                 insertNewPragma)
+import           Development.IDE.Types.Logger   (Priority (..), cmapWithPrio,
+                                                 logWith, (<+>))
+import           GHC.Generics                   (Generic)
+import           Ide.Plugin.RangeMap            (RangeMap)
+import qualified Ide.Plugin.RangeMap            as RangeMap
+import           Ide.PluginUtils                (getNormalizedFilePath,
+                                                 handleMaybeM, pluginResponse)
+import           Ide.Types                      (PluginDescriptor (..),
+                                                 PluginId (..),
+                                                 PluginMethodHandler,
+                                                 defaultPluginDescriptor,
+                                                 mkPluginHandler)
+import           Language.LSP.Types             (CodeAction (..),
+                                                 CodeActionKind (CodeActionRefactorRewrite),
+                                                 CodeActionParams (..), Command,
+                                                 List (..), Method (..),
+                                                 SMethod (..), TextEdit (..),
+                                                 WorkspaceEdit (WorkspaceEdit),
+                                                 fromNormalizedUri,
+                                                 normalizedFilePathToUri,
+                                                 type (|?) (InR))
+import qualified Language.LSP.Types.Lens        as L
+
+data Log
+  = LogShake Shake.Log
+  | LogCollectedRecordSelectors [RecordSelectors]
+  | LogRenderedRecordSelectors [ConvertedRecordSelector]
+
+instance Pretty Log where
+  pretty = \case
+    LogShake shakeLog -> pretty shakeLog
+    LogCollectedRecordSelectors recs -> "Collected record selectors:" <+> pretty recs
+    LogRenderedRecordSelectors recs -> "Rendered record selectors:" <+> pretty recs
+
+data CollectRecordSelectors = CollectRecordSelectors
+                    deriving (Eq, Show, Generic)
+
+instance Hashable CollectRecordSelectors
+instance NFData CollectRecordSelectors
+
+data CollectConvertedRecordSelectorsResult = CCRSR
+  { recordInfos       :: RangeMap ConvertedRecordSelector
```

Comment:
Could we have `RecordSelector`s here in the rule result instead? I ask because at the moment the rule unconditionally does the pretty-printing for all the record expressions in the file, which might be expensive. _Maybe_ laziness will save us, but it's simpler to just delay any pretty-printing work as late as possible. So ideally we could do it in the code action handler after we find the relevant record selector expressions. (Or eventually in a resolve method!)

Reply:
Yes, moving that computation now to the codeAction should tie in nicely with later implementing in in resolve.